autonomy/analyse/abci/app_spec.py

```diff
@@ -338,6 +343,51 @@ def abci_to_dfa(cls, abci_app_cls: Any, label: str = "") -> "DFA":
         )
 
 
+event_pattern = re.compile(r"Event\.(\w+)", re.DOTALL)
+
+
+def _check_unreferenced_events(abci_app_cls: AbciApp) -> None:
+    """Checks for unreferenced events in the AbciApp.
+
+    Checks that events defined in the AbciApp transition function are referenced
+    in the source code of the corresponding rounds or their superclasses. Note that
+    the function simply checks references in the "raw" source code of the rounds and
+    their (non builtin) superclasses. Therefore, it does not do any kind of static
+    analysis on the source code, nor checks for actual reachability of a return
+    statement returning such events.
+
+    :param abci_app_cls: AbciApp to check unreferenced events.
+    :raises DFASpecificationError: If there are unfererenced events in the AbciApp.
+    """
+
+    error_strings = []
+    timeout_events = {k.name for k in abci_app_cls.event_to_timeout.keys()}
+
+    for round_cls, round_transitions in abci_app_cls.transition_function.items():
+        trf_events = {
+            str(e).rsplit(".", 1)[1] for e in round_transitions
+        } - timeout_events
+        referenced_events = set()
+
+        for base in filter(
+            lambda x: x.__class__.__module__ != "builtins",
+            inspect.getmro(round_cls),
+        ):
+            src = textwrap.dedent(inspect.getsource(base))
+            referenced_events.update(event_pattern.findall(src))
```

Comment:
I disagree with this. Just the fact that the `src` contains an `Event.SOME_EVENT` does not mean that we also return this event from a round. For example, take a look at the APY's `APYEstimationAbstractRound` which is inherited by all the rounds:

https://github.com/valory-xyz/open-autonomy/blob/9ffbe860bc521842c882b33dd9bbebcac8747422/packages/valory/skills/apy_estimation_abci/rounds.py#L142-L164

Your script will require `NO_MAJORITY` and `FILE_ERROR` for all the rounds.

Reply:
A simple solution would be to remove these two helper methods, but the script would not be 100% correct.

Reply:
The problem is we can rely on static analysis only. We could make it a convention to adopt the policies checked by the script. It would keep the code easier to audit as well (don't have any code which isn't used in all inherited classes in the base class).

Reply:
Refactored APY in #1119, to remove code in the base class which isn't used in all inherited classes.